cmd/endpoint.go

```diff
@@ -0,0 +1,189 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/minio/minio-go/pkg/set"
+)
+
+// IsValidDistribution - checks whether given count is a valid distribution for erasure coding.
+func IsValidDistribution(count int) bool {
+	return (count >= 4 && count <= 16 && count%2 == 0)
+}
+
+// CheckLocalServerAddr - checks if serverAddr is valid and local host.
+func CheckLocalServerAddr(serverAddr string) error {
+	serverAddr = strings.TrimSpace(serverAddr)
+	if serverAddr == "" {
+		// empty server address means local host.
+		return nil
+	}
+
+	host, port, err := net.SplitHostPort(serverAddr)
+	if err != nil {
+		return err
+	}
+
+	// Check whether port is a valid port number.
+	p, err := strconv.Atoi(port)
+	if err != nil {
+		return fmt.Errorf("invalid port number in server address")
+	} else if p < 1 {
+		return fmt.Errorf("port number should be greater than zero in server address")
+	}
+
+	if host != "" {
+		localIPs := mustGetLocalIP4()
+		hostIPs, err := getHostIP4(host)
+		if err != nil {
+			return err
+		}
+
+		if localIPs.Intersection(hostIPs).IsEmpty() {
+			return fmt.Errorf("host in server address should be this server")
+		}
+	}
+
+	return nil
+}
+
+// IsEmptyPath - check whether given path is not empty.
+func IsEmptyPath(path string) bool {
+	return path == "" || path == "." || path == "/" || path == `\`
+}
+
+// EndpointType - enum for endpoint type.
+type EndpointType int
+
+const (
+	// UnknownEndpointType - unknown endpoint type enum.
+	UnknownEndpointType EndpointType = 0
+
+	// PathEndpointType - path style endpoint type enum.
+	PathEndpointType = 1
+
+	// URLEndpointType - URL style endpoint type enum.
+	URLEndpointType = 2
+)
+
+// Endpoint - any type of endpoint.
+type Endpoint struct {
+	Value      string
+	URL        *url.URL
+	IsDisabled bool
+	IsLocal    bool
+}
+
+// Type - returns type of endpoint.
+func (endpoint Endpoint) Type() EndpointType {
+	if IsEmptyPath(endpoint.Value) {
+		return UnknownEndpointType
```

Comment:
> In these cases, sometimes we would need to use URL.Scheme, URL.Opaque and/or URL.Path. This would make logical errors.

getPath() abstracts such details. But I agree that in case of "c:\path" storing "c" in u.Scheme does not look intuitive and also ep.Path is more readable than getPath(ep)

> In validation process, we know what are all endpoints are local. There is no need to use another function to do the same, hence IsLocal

isLocalStorage() will still be needed even if you have IsLocal, If we decide to have a separate Endpoint struct then we can have IsLocal too instead of calling isLocalStorage multiple times even during init process.

So a better Endpoint struct looks to be:
```
type Endpoint struct {
    path string
    url *url.URL
    isLocal bool
}
```
Either path will be valid or url will be valid depending on whether it represents a path or a url.  This will be more readable than having "Value" field which will be valid for both path and url ... plus Value is not useful as only url field gets used.